File: interveil/examples/typescript/basic-agent.ts
/**
 * Interveil — Basic TypeScript Agent Example
 *
 * This example shows how to wrap a simple agent with trace().
 * Run with: npx ts-node examples/typescript/basic-agent.ts
 */

import { trace, InterveillClient } from '../../src/index.js';

class SimpleAgent {
  async think(problem: string): Promise<string> {
    console.log(`[Agent] Thinking about: ${problem}`);
    await sleep(80);
    return `I should break this into steps: 1) Search for info, 2) Analyze results, 3) Summarize`;
  }

  async search(query: string): Promise<{ results: string[] }> {
    console.log(`[Agent] Searching: ${query}`);
    await sleep(340);
    return {
      results: [
        `Result 1: Information about ${query}`,
        `Result 2: More details on ${query}`,
        `Result 3: Related topic: ${query} advanced`,
      ],
    };
  }

  async analyze(data: unknown): Promise<string> {
    console.log(`[Agent] Analyzing results...`);
    await sleep(120);
    return `Analysis complete: Found 3 relevant results, highest confidence on result 1`;
  }

  async generateReport(analysis: string): Promise<string> {
    console.log(`[Agent] Generating report...`);
    await sleep(200);
    // Intentionally fail to demonstrate error highlighting
    if (Math.random() < 0.5) {
      throw new Error('Report generation failed: template not found');
    }
    return `Final report based on: ${analysis}`;
  }

  async summarize(report: string): Promise<string> {
    console.log(`[Agent] Summarizing...`);
    await sleep(60);
    return `Summary: ${report.slice(0, 100)}`;
  }
}

async function run() {
  const agent = new SimpleAgent();
  const tracedAgent = trace(agent, {
    sessionName: 'basic-agent-demo',
    agentId: 'simple-agent-v1',
    verbose: true,
    autoOpen: true,
  });

  try {
    const thought = await tracedAgent.think('How to improve developer productivity?');
    const searchResults = await tracedAgent.search('developer productivity tools 2025');
    const analysis = await tracedAgent.analyze(searchResults);
    const report = await tracedAgent.generateReport(analysis);
    const summary = await tracedAgent.summarize(report);
    console.log('\n[Agent] Done!', summary);
  } catch (err) {
    console.error('\n[Agent] Run failed (this is intentional for demo purposes):', (err as Error).message);
    console.log('[Agent] Check the Interveil UI to see the failure highlighted in red!');
  }
}

function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

run().catch(console.error);

File: interveil/examples/python/basic_agent.py
"""
Interveil — Basic Python Agent Example

Run with: python examples/python/basic_agent.py
(Requires the Interveil server to be running: interveil serve)

Demonstrates two styles:
  1. trace() one-liner wrapper — automatic tracing for any object
  2. InterveillClient manual API — fine-grained control over each event
"""

import sys
import time
import random
from datetime import datetime, timezone

sys.path.insert(0, "python")
from interveil import InterveillClient, trace, BlockedCommandError


def now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


# ── Style 1: trace() one-liner wrapper ───────────────────────────────────────

class ResearchAgent:
    """A simple agent that the trace() wrapper will instrument automatically."""

    def plan(self, goal: str) -> str:
        time.sleep(0.05)
        return f"Plan: search for '{goal}' then summarise."

    def search(self, query: str) -> dict:
        time.sleep(0.35)
        return {"results": [f"Result A for {query}", f"Result B for {query}"]}

    def summarise(self, results: dict) -> str:
        time.sleep(0.08)
        items = results.get("results", [])
        return "Summary: " + " | ".join(items)

    def run_shell(self, cmd) -> str:
        """Simulate a shell command (won't be reached for blocked patterns)."""
        return f"$ {cmd}\n(simulated output)"


def demo_trace_wrapper():
    print("\n=== Style 1: trace() one-liner wrapper ===")

    agent = trace(
        ResearchAgent(),
        session_name="trace-demo",
        verbose=True,
        # Any string patterns to block, on top of the built-in list
        blocked_patterns=["curl http://evil.com"],
    )

    goal = "latest advances in AI safety"

    plan    = agent.plan(goal)
    results = agent.search(goal)
    summary = agent.summarise(results)
    print(f"\nPlan:    {plan}")
    print(f"Results: {results}")
    print(f"Summary: {summary}")

    # Safe shell call — will trace normally
    output = agent.run_shell("ls -la /tmp")
    print(f"Shell:   {output}")

    # Dangerous call — object arg with buried 'rm -rf' caught before execution
    try:
        agent.run_shell({"command": "rm -rf /important/data"})
    except BlockedCommandError as e:
        print(f"\n[Blocked] {e}")

    agent.end()
    print("[trace] Session marked completed.")


# ── Style 2: InterveillClient manual API ─────────────────────────────────────

def demo_manual_api():
    print("\n=== Style 2: InterveillClient manual API ===")

    client = InterveillClient(port=3000)
    session = client.start_session(
        name="manual-api-demo",
        agent_id="python-agent-v1",
    )
    session_id = session["session_id"]
    print(f"[Agent] Session: {session_id}")

    step = 0

    start = time.time()
    time.sleep(0.08)
    client.emit(
        session_id=session_id, step_index=step, step_type="REASONING",
        input={"prompt": "How should I approach this research task?"},
        output={"thought": "Search → analyse → summarise"},
        timestamp=now(), duration_ms=int((time.time() - start) * 1000),
    )
    step += 1

    start = time.time()
    time.sleep(0.34)
    client.emit(
        session_id=session_id, step_index=step, step_type="TOOL_CALL",
        input={"tool": "web_search", "query": "AI agent frameworks 2025"},
        output={"results": ["LangChain", "AutoGen", "CrewAI", "Interveil"]},
        timestamp=now(), duration_ms=int((time.time() - start) * 1000),
    )
    step += 1

    start = time.time()
    time.sleep(0.85)
    client.emit(
        session_id=session_id, step_index=step, step_type="LLM_RESPONSE",
        input=None,
        output={"content": "LangChain is the most mature framework. AutoGen focuses on multi-agent conversations. Interveil provides observability across all of them."},
        timestamp=now(), duration_ms=int((time.time() - start) * 1000),
        model="gpt-4o",
        token_usage={"prompt_tokens": 87, "completion_tokens": 54, "total_tokens": 141},
    )
    step += 1

    if random.random() < 0.5:
        client.emit(
            session_id=session_id, step_index=step, step_type="ERROR",
            input={"action": "write_report", "path": "/tmp/report.md"},
            output=None, timestamp=now(), duration_ms=12,
            error={"message": "Permission denied", "code": "EACCES"},
        )
        step += 1
        client.end_session(session_id, "failed")
        print(f"[Agent] Done ({step} steps, failed). Open http://localhost:3000")
    else:
        client.end_session(session_id, "completed")
        print(f"[Agent] Done ({step} steps, completed). Open http://localhost:3000")


if __name__ == "__main__":
    demo_trace_wrapper()
    demo_manual_api()
